activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -44,21 +45,23 @@ def add_record(record); end
 
     class Transaction #:nodoc:
 
-      attr_reader :connection, :state, :records, :savepoint_name
+      attr_reader :connection, :state, :savepoint_name, :transactional_records
       attr_writer :joinable
 
       def initialize(connection, options)
         @connection = connection
         @state = TransactionState.new
-        @records = []
+        @transactional_records, @records = [], []
```

Comment:
good thought, but I would say `Set` would perform worse in this case, as the `push` method is slower on a Set and these list a pretty hot on insert.
Anyways, I think it is worth benchmarking and see some numbers, after I merge this, I will benchmark and see which one is better.thanks